File: java/src/main/java/banking/Account.java
package banking;

import fj.data.Either;
import fj.data.Option;

public class Account {

    private final BankService bankService;
    private final String id;

    public Account(BankService bankService,
                   String id) {
        this.bankService = bankService;
        this.id = id;
    }

    public Either<DepositError, AccountStatus> deposit(int cents) {
        throw new RuntimeException("TODO: not implemented yet");
    }

    public enum DepositError {
        CONNECTION_ERROR,
        ACCOUNT_NOT_FOUND,
        ACCOUNT_BLOCKED
    }

    public Either<WithdrawError, AccountStatus> withdraw(int cents) {
        throw new RuntimeException("TODO: not implemented yet");
    }

    public enum WithdrawError {
        CONNECTION_ERROR,
        ACCOUNT_NOT_FOUND,
        ACCOUNT_BLOCKED,
        NOT_ENOUGH_FUNDS
    }

    // Either<L, R> would have worked a bit better here
    // but we want to learn Option<A> as well, so… :)
    public Option<BankStatement> createStatement() {
        throw new RuntimeException("TODO: not implemented yet");
    }
}

File: kotlin/src/main/kotlin/banking/Account.kt
package banking

import arrow.core.Either
import arrow.core.Option

open class Account(private val bankService: BankService,
                   private val id: String) {

    open fun deposit(cents: Int): Either<DepositError, AccountStatus> {
        TODO("not implemented yet")
    }

    enum class DepositError {
        CONNECTION_ERROR,
        ACCOUNT_NOT_FOUND,
        ACCOUNT_BLOCKED
    }

    open fun withdraw(cents: Int): Either<WithdrawError, AccountStatus> {
        TODO("not implemented yet")
    }

    enum class WithdrawError {
        CONNECTION_ERROR,
        ACCOUNT_NOT_FOUND,
        ACCOUNT_BLOCKED,
        NOT_ENOUGH_FUNDS
    }

    // Either<L, R> would have worked a bit better here
    // but we want to learn Option<A> as well, so… :)
    open fun createStatement(): Option<BankStatement> {
        TODO("not implemented yet")
    }

}